iXBRLViewerPlugin/viewer/src/js/outline.js

```diff
@@ -0,0 +1,153 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import $ from 'jquery'
+
+// DocumentOutline chooses a fact for each presentation group (ELR) that
+// represents the start of that ELR.  This is done by deciding which ELRs each
+// fact participates in (see factInGroup()) and then finding the longest
+// continuous run of facts in document order that participate in each ELR.
+export function DocumentOutline(report) {
+    this._report = report;
+    var facts = report.facts().sort((a, b) => a.ixNode.docOrderindex - b.ixNode.docOrderindex);
+    var runLength = {};
+    var runStart = {};
+    var longestRun = {};
+    var longestRunStart = {};
+    this.dimensionMap = this._buildDimensionMap();
+    const elrs = report.relationshipGroups("pres");
+    for (const f of facts) {
+        if (f.isHidden()) {
+            continue;
+        }
+        for (const elr of elrs) {
+            if (this.factInGroup(f, elr)) {
+                if (!(elr in runLength)) {
+                    // Start counting a run
+                    runLength[elr] = 0;
+                    runStart[elr] = f;
+                }
+                runLength[elr]++;
+            }
+            else if (elr in runLength) { 
+                // End of a run
+                if (!(elr in longestRun) || longestRun[elr] < runLength[elr]) {
+                    longestRun[elr] = runLength[elr];
+                    longestRunStart[elr] = runStart[elr];
+                }
+                delete runLength[elr];
+                delete runStart[elr];
+            }
+        }
+    }
+
+    // End of document, check if any current runs are the longest run for the
+    // ELR.
+    for (const elr in runLength) {
+        if (!(elr in longestRun) || longestRun[elr] < runLength[elr]) {
+            longestRun[elr] = runLength[elr];
+            longestRunStart[elr] = runStart[elr];
+        }
+    }
+
+    this.sections = longestRunStart;
+}
+
+// Returns true if a fact participates in the given presentation group.
+DocumentOutline.prototype.factInGroup = function (fact, elr) {
+    // Roots are abstract so no need to check for concepts with outgoing
+    // relationships only.
+    if (this._report.getParentRelationshipsInGroup(fact.conceptName(), "pres", elr).length == 0) {
+        return false;
+    }
+    const fd = fact.dimensions();
+    const dm = this.dimensionMap[elr];
+    // Check all dimensions specified in this ELR
+    for (const [dim, spec] of Object.entries(dm)) {
+        // If a fact has a dimension, it must be in the list of permitted
+        // members, otherwise, the default member must be allowed
+        if ((dim in fd) ? !(fd[dim] in spec.members) : !spec.allowDefault) {
```

Comment:
What does this do for typed dimensions?

Reply:
Good question.  If a typed dimension is mentioned in the presentation then a fact would be required to have that dimension, with a value that corresponds to the QName of a concept that is a child of the typed dimension in the presentation.  The generation code doesn't cope with QName valued typed dimensions, so it would have to be a string that happened to correspond to the QName used internally by the viewer.

I think the current behaviour is harmless, but it's untested and we should probably handle typed dimensions explicitly.

I'm not sure what's normally done with typed dimensions in the presentation.  I assume that they're included with no children, if they're included at all, so we should probably make it so that if it's in the presentation, facts are required to have it.

Reply:
> I'm not sure what's normally done with typed dimensions in the presentation.  I assume that they're included with no children

I also don't know, but I sent a message to an internal channel to see if there's a convention, but my guess is the same as yours (I would be quite surprised if they're excluded altogether).

Reply:
I confirmed that's the convention.  As an example, the [Workiva 2020 10-K](https://www.sec.gov/ix?doc=/Archives/edgar/data/0001445305/000144530521000035/wk-20201231.htm) uses RevenueRemainingPerformanceObligationExpectedTimingOfSatisfactionStartDateAxis.

Reply:
Thanks - implemented in 261a1af and confirmed that it appears to work with the above document.